doc/tegu.8
.\"
.\" ---------------------------------------------------------------------------
.\"   Copyright (c) 2013-2015 AT&T Intellectual Property
.\"
.\"   Licensed under the Apache License, Version 2.0 (the "License");
.\"   you may not use this file except in compliance with the License.
.\"   You may obtain a copy of the License at:
.\"
.\"       http://www.apache.org/licenses/LICENSE-2.0
.\"
.\"   Unless required by applicable law or agreed to in writing, software
.\"   distributed under the License is distributed on an "AS IS" BASIS,
.\"   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\"   See the License for the specific language governing permissions and
.\"   limitations under the License.
.\" ---------------------------------------------------------------------------
.\"

.\"
.\"		tegu Manual Page
.\"
.\"     Date:		04 Jul 2015
.\"		Author:		Robert Eby (eby@research.att.com)
.\"
.\"     Mods:		04 Jul 2015 - Created
.\"					16 Aug 2015 - Finished descriptions.
.\"					01 Sep 2015 - Add section about state mismatch.
.\"					24 Nov 2015 - Add options to add-mirror
.\"
.TH TEGU 8 "Tegu Manual"
.CM 4
.SH NAME
tegu \- the main Tegu process
.SH SYNOPSIS
\fBtegu\fP [\fB-C config-file\fP] [\fB-c checkpoint_file\fP] [\fB-f host[:port]\fP] [\fB-p port\fP] [\fB-s admin-cookie\fP] [\fB-v\fP]

.SH DESCRIPTION
\fItegu(8)\fR is the main process in the Tegu system.
It provides the HTTP API interface,
maintains a view of the network,
communicates with Openstack to determine tenants, ports, etc.,
communicates with Floodlight (if configured) to determine switch topology,
communicates with agents that distribute commands to other physical nodes (via \fIssh\fP),
and keeps track of scheduling of reservations and other operations.
.PP
The API interface provided by Tegu is described in the API INTERFACE section below.
The HTTP server may require that its clients use SSL, if so configured.
.PP
\fItegu\fR is not started directly by the tegu init script.
Rather, the service init script starts a Python program called \fItegu_ha\fP which will
start tegu on exactly one host.
The tegu_ha script runs on each tegu node in a continuous loop checking for hearbeats from
all the tegu nodes found in the /etc/tegu/standby_list once every 5 secs (default).
A heartbeat is obtained by invoking the "tegu ping" command via the HTTP API.
If exactly one tegu is running, the script does nothing.
If no tegu is running, then the local tegu is activated after waiting for 5*priority
seconds to let a higher priority tegu take over first.
A tegu node's priority is determined by its order in the standby_list file.
If the current node's tegu is running and another tegu is found, then a conflict
resolution process is invoked whereby the last modified timestamps from checkpoint files
from both tegu's are compared, and the tegu with the older checkpoint is deactivated.
The rationale is that the F5 only keeps a single tegu active at a time,
so the tegu with the most recent checkpoint ought to be the active one.
.PP
It is important to note, that many of the commands that list Tegu's state (such as
\fIlist-mirrors\fP, for example) will usually list the state of Tegu's internal database,
and will not go out to query the actual network.
Assuming no rogue actors (system crashes, other misbehaving software, or system administrator actions),
Tegu's state should normally match the state of the network.
And of course, reservations in Tegu may have expired or not yet commenced, in which case
they will not be apparent in the network.
However, if the state of the network does not match Tegu's internal state (because of rogue
actors), the first thing to try is to restart Tegu.
This will have the effect of re-pushing all reservations out to the network in order to
resynchronize the network's state with Tegu's state.
.SH COMMAND LINE OPTIONS
\fITegu\fR interprets the following options:
.\" ==========
.TP 8
.B \-C config-file
Configure tegu using \fIconfig-file\fP.
The structure and contents of the configuration file are described in \fItegu.cfg(5)\fP.
.\" ==========
.TP 8
.B \-c checkpoint_file
Specifies a checkpoint file that Tegu should initialize from.

.\" ==========
.TP 8
.B \-f host[:port]
Specifies the Floodlight host and port to use.
This can also be specified via the configuration file.
.\" ==========
.TP 8
.B \-p port
Causes Tegu to start an HTTP server on \fIport\fP.
The default port is 29444.
The decision as to whether this is an HTTP or HTTPS server is determined by the presence
or absence of a private key and certificate in the configuration file.
.\" ==========
.TP 8
.B \-s admin-cookie
Set the "super cookie" used by Tegu to \fIadmin-cookie\fP.
This cookie can be used anywhere a cookie must be provided to affect a reservation.
If not set, a pre-determined super-cookie is used (compiled into the code).
.TP 8
.B \-v
Turns on verbose logging (changes the logging level to 1).

.SH OPERATIONS
Tegu is packaged as a .deb file for installation on Ubuntu Linux.
It can be started and stopped as the \fItegu\fP service using the standard Linux
service facility; e.g.
.IP
\f(CWservice tegu start\fP
.P
In normal operation, Tegu runs as a number of processes under the \fItegu\fP login;
a Python script \fItegu_ha\fP  which makes sure that Tegu is running, and fails over to a backup server if need be,
a \fItegu\fP process which provides the main functionality of the service,
and a number (usually 5) of \fItegu_agent\fP processes which pass commands from
Tegu to other physical hosts in a network of machines (via \fIssh\fP).

.SH "API INTERFACE"
Commands to Tegu from a client such as \fItegu_req\fP are packaged as HTTP requests to
the Tegu HTTP server.
For the most part, these are sent as POSTs to the URL /tegu/api (however,
mirroring commands follow a more ReST-ful paradigm; see below).
The body of the POST should have a Content-type of text, and contains a command line
that lists the command and arguments.
Authentication, if required, is provided via a prefix of \f(CWauth=token\fP
containing a Keystone token that is used to verify the user.
The response to each Tegu request is a JSON body containing any requested information as
well as error status.
.P
If Tegu has just started and is still not accepting commands,
then most commands (with the exception of \fIping\fP and \fIverbose\fP) will fail.
The error message that is returned will indicate this.

.SS Bandwidth Allocation Commands
.TP 8
.B reserve [bandwidth_in,]bandwidth_out [start-]expiry host1-host2 cookie dscp
Makes a bandwidth reservation.
.TP 8
.B [auth=token] reservation reservation-id [cookie]
This command is issued as a DELETE, not a POST.
This caused the named reservation to be cancelled.
The cookie is required if the initial reservation was made with a cookie, and must
match either the initial reservation's cookie or the super cookie.
.TP 8
.B [auth=token] setdiscount value
Sets the discount value to either a percentage, or an absolute value.
See the description of \fIsetdiscount\fP on tegu_req(1) for details of the value.
.TP 8
.B [auth=token] setulcap tenant percentage
Sets the user link capacity for a tenant.
\fIpercentage\fP is a number from 0 to 100.
.TP 8
.B [auth=token] listulcap
List all user link capacities known by the network manager.
.TP 8
.B [auth=token] listres
List all reservations (pledges) that Tegu knows about.
.TP 8
.B [auth=token] qdump
This is the API equivalent of the \fItegu_req listqueue\fP command.
It returns a JSON list of all queues on the switches or bridges being managed.
.SS Topology Commands
.TP 8
.B [auth=token] graph [key=value ...]
This returns a JSON network graph, describing all of the switches in the network.
.TP 8
.B [auth=token] listhosts [key=value ...]
Returns a JSON list of hosts which includes IP address, name, switch(es) and port(s).
.SS Flow Steering Commands
.TP 8
.B steer {[start-]end|+seconds} tenant src-host dest-host mbox-list cookie
This is a prototype flow-steering command (deprecated).

.SS Mirroring Commands
Mirroring commands follow a ReST-ful paradigm, so this section is a little bit different.
All mirroring URL's are located under /tegu/mirrors/.
Mirroring commands that expect input should receive a JSON string in the content body,
with the Content-type set to \f(CWapplication/json\fP, although Tegu will not (currently)
complain if the Content-type header is wrong.
In addition, the authorization token for mirroring requests is passed in the
\fBX-Auth-Tegu\fP header, rather than in the content body.
.TP 8
.B add-mirror [start-]end port1[,port2...] output [cookie] [vlan]
To add a mirror, a POST is done to the URL /tegu/mirrors.
The body of the post should look like:
.IP
.nf
.ft CW
{
	"start_time": "nnn",                 // optional
	"end_time": "nnn",                   // required
	"output": "<output spec>",           // required
	"port": [ "port1" , "port2", ...],   // required
	"vlan": "vlan",                      // optional
	"cookie": "value",                   // optional
	"name": "mirrorname",                // optional
	"options": "optionlist",             // optional
}
.ft P
.fi
See tegu_req(1) for the meanings of the fields.
.IP
Because multiple mirrors may be created as a result of one call,
we return a JSON array of results, one for each mirror:
.IP
.nf
.ft CW
[
	{
		"name": "mirrorname",   // tegu or user-defined mirror name
		"url": "url",           // URL to use for DELETE or GET
		"error": "err"          // error message (if any)
	},
	....
]
.ft P
.fi
.TP 8
.B del-mirror name [cookie]
To delete a mirror, a DELETE request is sent to the URL
.IP
/tegu/mirrors/\fImirrorname\fP/[?cookie=\fIcookie\fP]
.IP
If the named mirror exists, and the cookie (if any) matches,
the mirror is deleted and a 204 (No Content) is returned.
Other possible return codes are 404 (Not Found) and 401 (Unauthorized).
.TP 8
.B list-mirrors
To list all mirrors that Tegu knows about, a GET of /tegu/mirrors/ is performed.
If authorized, a JSON array is returned, listing mirror names and URLs.
.IP
.nf
.ft CW
[
	{
		"name": "mir-5599a59a_0",                              // tegu or user-defined mirror name
		"url": "http://localhost:29444/tegu/mirrors/mir-5599a59a_0/"  // URL to use for DELETE or GET
	},
	....
]
.ft P
.fi
.TP 8
.B show-mirror name [cookie]
To show details of a particular mirror, a GET request is sent to the URL
.IP
/tegu/mirrors/\fImirrorname\fP/[?cookie=\fIcookie\fP]
.IP
If the named mirror exists, and the cookie (if any) matches, a JSON response is returned
describing the mirror.
Other possible return codes are 404 (Not Found) and 401 (Unauthorized).
.IP
.nf
.ft CW
{
	"name": "mir-5599a59a_0",
	"start_time": 1436132762,
	"end_time": 1436134567,
	"port": [
		"fa:16:3e:5f:df:60"
	],
	"output": "135.25.85.99",
	"physical_host": "mtmac3",
	"pushed": true,
	"paused": false,
	"pending": false,
	"active": true,
	"expired": false,
	"url": "http://localhost:29444/tegu/mirrors/mir-5599a59a_0/"
}
.ft P
.fi

.SS Miscellaneous Commands
.TP 8
.B ping
This command is used to check connectivity to the Tegu system itself.
It will always return success, so if a successful connection is made, and a
response received, the client can assume Tegu is running.

.TP 8
.B [auth=token] refresh [ vmname | tenant/vmname ]
When a user migrates a VM they are expected to cancel and reissue reservation requests before
and after the migration of any VM that is a apart of an existing reservation (whether future
or active).
The operations staff may issue a privileged command which refreshes any existing reservations
for a VM that has already been migrated.
Following the migration of a VM, the refresh command will reset reservations that were associated
with the VM, and will preserve reservation related information (name and cookie) such that
the reservation owner can still manage the reservation.
The following syntax illustrates how the tegu_req command line interface can be used to
refresh the reservations associated with the VM named \fIdemovm8\fP in the \fIdemo\fP tenant:

tegu_req refresh demo/demovm8

.TP 8
.B verbose level [subsystem]
This is the API version of the \fItegu_req\fP verbose command.
Note that no authentication token is required.
For details on valid values for \fIsubsystem\fP, see the tegu_req(1) manual page.

.TP 8
.B [auth=token] listconns [name]
Returns a JSON description of the switches and ports for the named host.

.SH FILES
.TP 15
/var/lib/tegu
Normal directory for Tegu checkpoints.
.TP 15
/var/log/tegu
Normal directory for Tegu logfiles.
.TP
/etc/tegu/tegu.cfg
The Tegu configuration file.
.TP
/etc/tegu/phys_net_static.json
The physical network description.

.SH SEE ALSO
ssh(1), tegu_req(1), rjprt(1), tegu.cfg(5), service(8)


main/tegu.go
// vi: sw=4 ts=4:
/*
 ---------------------------------------------------------------------------
   Copyright (c) 2013-2015 AT&T Intellectual Property

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at:

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 ---------------------------------------------------------------------------
*/


/*

	Mnemonic:	tegu
	Abstract:	The middle layer that sits between the cQoS and the openflow controller (Skoogi)
				providing an API that allows for the establishment and removal of network
				reservations.

				Command line flags:
					-C config	-- config file that provides openstack credentials and maybe more
					-c chkpt	-- checkpoint file, last set of reservations
					-f host:port -- floodlight (SDNC) host:port
					-p port		-- tegu listen port (4444)
					-s cookie	-- super cookie
					-v			-- verbose mode

	Date:		20 November 2013
	Author:		E. Scott Daniels

	Mods:		20 Jan 2014 : Added support to allow a single VM in a reservation (VMname,any)
							+nnn time now supported on reservation request.
				10 Mar 2014 : Converted to per-path queue setting (ingress/egress/middle queues).
				13 Mar 2014 : Corrected 'bug' with setting pledges where both hosts connect to the
							same switch. (bug was that it wasn't yet implemented.)
				03 Apr 2014 : Added endpoint support for reservations and flowmods.
				05 May 2014 : Changes to support merging gateways into the graph and for receiving
							responses back from the agent.
				13 May 2014 : Changes to support dscp-exit value supplied on reservation.
				16 May 2014 : Corrected bug with specifying the "exit" dscp value.
				18 May 2014 : Now supports cross tenant reservations.
				29 May 2014 : Now supports default openstack value propagation in config file.
				05 Jun 2014 : Added pause/resume ability.
				06 Jun 2014 : Added TLS/SSL support.
				09 Jun 2014 : Added token authorisation support for reservations.
				11 Jun 2014 : All paths option added-- reservation will be based on capacity over all
							paths between h1 and h2.
				16 Jun 2014 : Added ability to authorise privileged commands with a token generated using
							the default (admin) user name given in the config file.
				17 Jun 2014 : Added support for transport ports on reservation host names.
				25 Jun 2014 : Added user (project, tenant, whatever) level caps on each link (gak).
				07 Jul 2014 : Added refresh API request.
				15 Jul 2014 : Added support for partial reservation path when only one endpoint is given a
							valid token on the reservation request.
				21 Jul 2014 : Fixed bug -- checkpoint not including user link caps
				29 Jul 2014 : Added mlag support.
				13 Aug 2014 : Changes to include network hosts in the list of hosts (incorporate library
							changes)
				15 Aug 2014 : Bug fix (201) and stack dump fix (nil ptr in osif).
				19 Aug 2014 : Bug fix (208) prevent duplicate tegu's from running on same host/port.
				20 Aug 2014 : Bug fix (210) shift dscp values properly in the match part of a flowmod.
				22 Aug 2014 : Added protocol support to steering. (steer)
				27 Aug 2014 : Fq-mgr changes to allow for metadata checking on flowmods.
				30 Aug 2014 : Pick up bug fix made to ostack library.
				03 Sep 2014 : Corrected bug in resmgr/fqmgr introduced with 27 Aug changes (transport type ignored).
				05 Sep 2014 : Tweak to link add late binding port to pick up the lbp when port < 0 rather than 0.
				08 Sep 2014 : Corrected bugs with tcp oriented proto steering (dest based). (steer)
				23 Sep 2014 : Support for rate limiting bridge
				29 Sep 2014 : Nil pointer exception (bug #216) corrected (gizmo change)
				30 Sep 2014 : Deal with odd hostnames that were being returned by ccp's version of openstack.
				09 Oct 2014 : Bug fix (228) -- don't checkpoint until all initialised.
				14 Oct 2014 : Rebuild to pick up library changes that get network 'hosts'
							as host only where OVS is running and not all network hosts.
				19 Oct 2014 : Added bidirectional bandwidth support (bug 228). (version bump to 3.0.1 because of
							extra testing needed.)
				23 Oct 2014 : Added better diagnostics to the user regarding capacity rejection of reservation (bug 239)
				29 Oct 2014 : Corrected issue where vlan id was being set when both VMs are on the same switch (bug 242)
				30 Oct 2014 : Corrected bug with setting the source/dest flag for external IP addresses in flowmod req (bug 243)
				03 Nov 2014 : Merged steering and lite branches into steering branch; version bump to 3.2.
				04 Nov 2014 : Build to pick up ostack library change.
				10 Nov 2014 : Build to pick up ostack library change (small tokens).
				11 Nov 2014 : Change to support host name suffix in fqmgr.
				12 Nov 2014 : Change to strip phys host suffix from phys map.
				13 Nov 2014 : Correct out of bounds exception in fq-manager.
				17 Nov 2014 : Converted the openstack interface to a lazy update method rather than attempting to
							prefetch all of the various translation maps and then to keep them up to date.
				19 Nov 2014 : Correct bug in checkpoint path attachment to reservation.
				24 Nov 2014 : Floating IP address requirement for cross tenant reservations, and reservations between VM and an
							 external host has been modified.
				04 Dec 2014 : Uses enabled host list from ostack interface.
				05 Dec 2014 : Added work round for keystone/privs token investigation issue in AIC once they started using LDAP.
				07 Dec 2014 : Corrected default tickler time for set intermediate queues.
				16 Dec 2014 : Corrected bugs in lazy update translation of VM ID to ip.
				08 Jan 2015 : Corrected a bug that was causing the wrong gateway to be selected as the endpoint for an
							external reservation.
				09 Jan 2015 : Augmented the 1/8 fix to not assume the subnet list is limited by project in the creds.
							Changes to allow for finding all IP addresses assigned to a VM, not just the first.
				16 Jan 2015 : Added support for mask on the transport port specification.
							Changed meta table flow-mod generation to happen with queue changes and to push only to nodes
							in the queue list rather than to all nodes.
				26 Jan 2015 : Corrected bug in fq_mgr that was causing inbound data to use meta marking in base+1 rather than base table.
				27 Jan 2015 : Allow bandwidth to be specified as a decimal (e.g 10.2M) on a reservation command.
				29 Jan 2015 : Changes to send fmod requests to the ssh-broker enabled agent.
				01 Feb 2015 : Corrected bug introduced when host name removed from fmod command (agents with ssh-broker change).
				09 Feb 2015 : Added work round to deal with OVS/openflow small hard timeout limit.
				10 Feb 2015 : Corrected bug with refresh.
				18 Feb 2015 : Corrected bug with slice index and enabled both host name types on create queue list.
				24 Feb 2015 : Added mirroring (version => 3.1.3).
				11 Mar 2015 : Corrected bug in network manager causing coredump when gateway info missing on bleat message.
				25 Mar 2015 : Added support for IPv6 in bandwidth reservations, and new bandwidth flow-mod script in the agent.
				31 Mar 2015 : Added ability to force a load of any specific project's VM data into the graph, or all that are
							known to Tegu.
				03 Apr 2015 : Added role verification for graph and listhosts to allow any admin to submit these from afar.
				10 Apr 2015 : Added ability to delete reservation via post, corrected slice bounds error (tokeniser library fix).
				16 Apr 2015 : Now supports openstack regions.
				20 Apr 2015 : Corrected bug -- not passing external address direction to agent.
				18 May 2015 : Added discount.
				26 May 2015 : Converted to using the new pledge as an interface.
				01 Jun 2015 : Added reservation dup checking.
				04 Jun 2015 : Added authentication to mirror request processing.
				09 Jun 2015 : Added oneway bandwidth reservations.
				16 Jun 2015 : Extended oneway bandwidth reservations to include queues.
				25 Jun 2015 : Corrected bug with mirror expiration and xlation of proj/vm to mac.
				02 Jul 2015 : Corrected physical host fetch bug in osif.
				06 Jul 2015 : Version bump
				29 Jul 2015 : Tracker bug fixes (263,266) version bump.
				03 Sep 2015 : Correct panic in network.go.
				08 Sep 2015 : Prevent checkpoint files with same timestamp (gh#22).
				08 Oct 2015 : Correct bug causing all reservations to be pushed when reservation change state happens.
				09 Oct 2015 : Use 'admin' project for default phys host list.
				16 Nov 2015 : Add restart message to logfile.
				09 Dec 2015 : Require token on requests to always be token/project.

	Version number "logic":
				3.0		- QoS-Lite version of Tegu
				3.0.1	- QoS-Lite version of Tegu with lazy openstack information gathering (17 Nov 2014)
				3.1		- QoS-Lite with steering and mirroring added
				3.2		- QoS-Lite with steering and WACC support added
	Trivia:		http://en.wikipedia.org/wiki/Tupinambis
*/

package main

import (
	"flag"
	"fmt"
	"os"
	"sync"
	"time"

	"github.com/att/gopkgs/bleater"
	"github.com/att/gopkgs/ipc"
	"github.com/att/tegu/managers"
)

var (
	sheep *bleater.Bleater
)

func usage( version string ) {
	fmt.Fprintf( os.Stdout, "tegu %s\n", version )
	fmt.Fprintf( os.Stdout, "usage: tegu [-C config-file] [-c ckpt-file] [-f floodlight-host] [-p api-port] [-s super-cookie] [-v]\n" )
}

func main() {
	var (
		version		string = "v3.1.5/1c105"		// 3.1.x == steering branch version (.2 steering only, .3 steering+mirror+lite)
		cfg_file	*string  = nil
		api_port	*string						// command line option vars must be pointers
		verbose 	*bool
		needs_help 	*bool
		fl_host		*string
		super_cookie *string
		chkpt_file	*string

		// various comm channels for threads -- we declare them here so they can be passed to managers that need them
		nw_ch	chan *ipc.Chmsg		// network graph manager
		rmgr_ch	chan *ipc.Chmsg		// reservation manager
		osif_ch chan *ipc.Chmsg		// openstack interface
		fq_ch chan *ipc.Chmsg		// flow queue manager
		am_ch chan *ipc.Chmsg		// agent manager channel

		wgroup	sync.WaitGroup
	)

	sheep = bleater.Mk_bleater( 1, os.Stderr )
	sheep.Set_prefix( "main/3.1" )

	needs_help = flag.Bool( "?", false, "show usage" )

	chkpt_file = flag.String( "c", "", "check-point-file" )
	cfg_file = flag.String( "C", "", "configuration-file" )
	fl_host = flag.String( "f", "", "floodlight_host:port" )
	api_port = flag.String( "p", "29444", "api_port" )
	super_cookie = flag.String( "s", "", "admin-cookie" )
	verbose = flag.Bool( "v", false, "verbose" )

	flag.Parse()									// actually parse the commandline

	if *needs_help {
		usage( version )
		os.Exit( 0 )
	}

	if( *verbose ) {
		sheep.Set_level( 1 )
	}
	sheep.Baa( 1, "tegu %s started", version )
	sheep.Baa( 1, "http api is listening on: %s", *api_port )

	if *super_cookie == "" {							// must have something and if not supplied this is probably not guessable without the code
		x := "20030217"
		super_cookie = &x
	}

	nw_ch = make( chan *ipc.Chmsg, 128 )					// create the channels that the threads will listen to
	fq_ch = make( chan *ipc.Chmsg, 1024 )			// reqmgr will spew requests expecting a response (asynch) only if there is an error, so channel must be buffered
	am_ch = make( chan *ipc.Chmsg, 1024 )			// agent manager channel
	rmgr_ch = make( chan *ipc.Chmsg, 1024 );			// buffered to allow fq to send errors; should be more than fq buffer size to prevent deadlock
	osif_ch = make( chan *ipc.Chmsg, 1024 )

	err := managers.Initialise( cfg_file, &version, nw_ch, rmgr_ch, osif_ch, fq_ch, am_ch )		// specific things that must be initialised with data from main so init() doesn't work
	if err != nil {
		sheep.Baa( 0, "ERR: unable to initialise: %s\n", err );
		os.Exit( 1 )
	}
	managers.Log_Restart( version )

	go managers.Http_api( api_port, nw_ch, rmgr_ch )				// start early so we bind to port quickly, but don't allow requests until late
	go managers.Res_manager( rmgr_ch, super_cookie ); 				// manage the reservation inventory
	go managers.Osif_mgr( osif_ch )									// openstack interface; early so we get a list of stuff before we start network
	go managers.Network_mgr( nw_ch, fl_host )						// manage the network graph
	go managers.Agent_mgr( am_ch )
	go managers.Fq_mgr( fq_ch, fl_host );

	my_chan := make( chan *ipc.Chmsg )								// channel and request block to ping net, and then to send all sys up
	req := ipc.Mk_chmsg( )

	/*
		Block until the network is initialised. We need to do this so that when the checkpoint file is read reservations
		can be added without missing network pieces.  Even if there is no checkpoint file, or it's empty, blocking
		prevents reservation rejections because the network graph isn't in working order.  At the moment, with lazy
		updating, the block is until we have a physical host map back from the agent world.  This can sometimes take
		a minute or two.
	*/
	for {																	// hard block to wait on network readyness
		req.Response_data = 0
		req.Send_req( nw_ch, my_chan, managers.REQ_STATE, nil, nil )		// 'ping' network manager; it will respond after initial build
		req = <- my_chan													// block until we have a response back

		if req.Response_data.(int) == 2 {									// wait until we have everything that the network needs to build a reservation
			break
		}

		sheep.Baa( 2, "waiting for network to initialise: need state 2, current state = %d", req.Response_data.(int)  )
		time.Sleep( 5 * time.Second )
	}

	if *chkpt_file != "" {
		sheep.Baa( 1, "network initialised, sending chkpt load request" )
		req.Send_req( rmgr_ch, my_chan, managers.REQ_LOAD, chkpt_file, nil )
		req = <- my_chan												// block until the file is loaded

		if req.State != nil {
			sheep.Baa( 0, "ERR: unable to load checkpoint file: %s: %s\n", *chkpt_file, req.State )
			os.Exit( 1 )
		}
	} else {
		sheep.Baa( 1, "network initialised, opening up system for all requests" )
	}

	req.Send_req( rmgr_ch, nil, managers.REQ_ALLUP, nil, nil )		// send all clear to the managers that need to know
	managers.Set_accept_state( true )								// http doesn't have a control loop like others, so needs this

	wgroup.Add( 1 )					// forces us to block forever since no goroutine gets the group to dec when finished (they dont!)
	wgroup.Wait( )
	os.Exit( 0 )
}
